File: scm-csharp/scheme/primitives/PrimitiveHashTableComparator.cs
namespace scheme;

public class PrimitiveHashTableComparator : Primitive
{
    public override string Name() => "%hash-table-comparator";
    public override string Info() => "";

    public override object Apply(SourcePos? pos, object[] arguments)
    {
        CheckArgs(pos, arguments, 1, 1);
        if (arguments[0] is not SchemeHashTable ht)
            throw new SchemeError(pos, "%hash-table-comparator: not a hash table");
        return ht.Comparator ?? Value.F;
    }
}

File: scm-java/src/main/scheme/primitives/PrimitiveHashTableComparator.java
package scheme.primitives;

import scheme.*;

public class PrimitiveHashTableComparator extends Primitive {
    @Override
    public String name() { return "%hash-table-comparator"; }

    @Override
    public String info() { return ""; }

    @Override
    public Object apply(SourcePos pos, Object[] arguments) {
        checkArgs(pos, arguments, 1, 1);
        if (!(arguments[0] instanceof SchemeHashTable))
            throw new SchemeError(pos, name() + ": not a hash table");
        SchemeHashTable ht = (SchemeHashTable) arguments[0];
        return ht.comparator != null ? ht.comparator : Value.F;
    }
}
